cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/SourceControlManager.java

```diff
@@ -0,0 +1,244 @@
+  }
+
+  // Deletes a directory by recursively deleting its contents.
+  private boolean deleteDirectory(File dir) {
+    File[] allContents = dir.listFiles();
+    if (allContents != null) {
+      for (File file : allContents) {
+        deleteDirectory(file);
+      }
+    }
+    return dir.delete();
+  }
+
+  private <C extends TransportCommand> C getAuthenticatedCommandWithTimeout(C command) throws AuthenticationException {
+    command.setCredentialsProvider(authStrategy.getCredentialProvider(context));
+    Optional<TransportConfigCallback> transport = authStrategy.getTransportConfigCallback(context);
+    transport.ifPresent(command::setTransportConfigCallback);
+    command.setTimeout(gitCommandTimeoutSeconds);
+    return command;
+  }
+
+  /**
+   * Initialises the Git repository by cloning remote.
+   *
+   * @throws GitAPIException         when a Git operation fails.
+   * @throws AuthenticationException when correct authentication credentials are not available.
+   * @throws IOException             when file or network I/O fails.
+   */
+  public void initialise() throws GitAPIException, AuthenticationException, IOException {
+    if (git != null) {
+      return;
+    }
+    // Clean up the directory if it already exists.
+    if (localRepoDirectory.exists() && !deleteDirectory(localRepoDirectory)) {
+      throw new IOException("Failed to clean up local repository directory " + context.getRepoBasePath());
+    }
+    String branch = "refs/heads/" + context.getRepositoryConfig().getDefaultBranch();
+    CloneCommand command = Git.cloneRepository()
+      .setURI(context.getRepositoryConfig().getDefaultBranch())
+      .setDirectory(localRepoDirectory)
+      .setBranchesToClone(Collections.singleton((branch)))
+      .setBranch(branch);
+
+    command = getAuthenticatedCommandWithTimeout(command);
+    command.call();
+  }
+
+  /**
+   * Returns an {@link AuthStrategy} for the given Git repository provider and auth type.
+   *
+   * @return an instance of {@link  AuthStrategy}.
+   * @throws AuthStrategyNotFoundException when the corresponding {@link AuthStrategy} is not found.
+   */
+  private AuthStrategy getAuthStrategy() throws AuthStrategyNotFoundException {
+    RepositoryConfig config = this.context.getRepositoryConfig();
+    if (config.getProvider() == Provider.GITHUB) {
+      if (config.getAuth().getType() == AuthType.PAT) {
+        return new GitPATAuthStrategy(this.context.getStore());
+      }
+    }
+    throw new AuthStrategyNotFoundException(String.format("No strategy found for provider %s and type %s.",
+                                                          config.getProvider(),
+                                                          config.getAuth().getType()));
+  }
+
+  /**
+   * Validated the configuration provided to this manager.
+   *
+   * @return a {@link List} of {@link  Exception}s occurred during validation.
+   */
+  public List<Exception> validateConfig() {
```

Comment:
Could we use `RepositoryValidationFailure`? We can make `RepositoryValidationFailure` extend `Exception` if we need exceptions

Reply:
The idea is to re-use `InvalidRepositoryConfigException` and make the PUT endpoint return consistent errors response